telemetry/app.go

```diff
@@ -0,0 +1,135 @@
+package telemetry
+
+import (
+	"context"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+
+	"cosmossdk.io/log"
+	storetypes "cosmossdk.io/store/types"
+	abci "github.com/cometbft/cometbft/abci/types"
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/mempool"
+	"github.com/hashicorp/go-metrics"
+)
+
+const (
+	blockTxsSizeMetric    = "block_txs_size"
+	blockResultSizeMetric = "block_result_size"
+	diskUsageMetric       = "disk_usage"
+)
+
+// InitAppMetrics initializes the block specific metrics for the application.
+func InitAppMetrics(homeDir string) func(*baseapp.BaseApp) {
+	return func(app *baseapp.BaseApp) {
+		app.SetPrepareProposal(initPrepareProposalHandlerWithMetrics(app))
+		app.SetStreamingManager(initStreamingManagerWithMetrics(app, path.Join(homeDir, "data")))
+	}
+}
+
+// initPrepareProposalHandlerWithMetrics initializes the prepare proposal handler
+// with the app metrics.
+// It gathers the block txs size to emit them as a gauge metric.
+func initPrepareProposalHandlerWithMetrics(app *baseapp.BaseApp) sdk.PrepareProposalHandler {
+	// Create a NoOpMempool for the application and get the default prepare proposal
+	// handler as per NewBaseApp implementation.
+	// See https://github.com/cosmos/cosmos-sdk/blob/v0.50.4/baseapp/baseapp.go#L221
+	app.SetMempool(mempool.NoOpMempool{})
```

Comment:
Are we still going to have a mempool? It's critical for the operation of the node.

Reply:
We are already using the no-op mempool anyway. It seems that it's possible to remove this setting allowing the application to freely set its own upstream if needed.